src/main/java/net/minecraftforge/common/util/ChunkedTileEntityList.java

```diff
@@ -0,0 +1,369 @@
+        if (list != null)
+            return list.contains(te);
+        return false;
+    }
+
+    public void add(TileEntity... tileEntities)
+    {
+        for (int i = 0; i < tileEntities.length; i++)
+        {
+            add(tileEntities[i]);
+        }
+    }
+
+    @Override
+    public boolean addAll(Collection<? extends TileEntity> tileEntities)
+    {
+        for (TileEntity tileEntity : tileEntities)
+        {
+            add(tileEntity);
+        }
+        return true;
+    }
+
+    @Override
+    public boolean add(TileEntity te)
+    {
+        ChunkPos pos = getChunkPos(te);
+        List<TileEntity> list = getChunkTEList(pos);
+        if (list == null)
+        {
+            list = new ArrayList<>();
+            list.add(te);
+            content.put(pos, list);
+            return true;
+        }
+        else
+            return list.add(te);
+    }
+
+    @Override
+    public void add(int index, TileEntity te)
+    {
+        add(te);
+    }
+
+    @Override
+    public boolean addAll(int index, Collection<? extends TileEntity> tileEntities)
+    {
+        return addAll(tileEntities);
+    }
+
+    @Override
+    public boolean removeAll(Collection<?> tileEntities)
+    {
+        boolean removedAll = true;
+        for (Object object : tileEntities)
+        {
+            if (!remove(object))
+                removedAll = false;
+        }
+        return removedAll;
+    }
+
+    @Override
+    public boolean remove(Object te)
+    {
+        if (!(te instanceof TileEntity))
+            return false;
+        ChunkPos pos = getChunkPos((TileEntity) te);
+        List<TileEntity> values = getChunkTEList(pos);
+        if (values != null)
+            if (values.remove(te))
+            {
+                if (values.isEmpty())
+                    removeChunk(pos);
+                return true;
+            }
+        return false;
+    }
+
+    private boolean removeChunk(ChunkPos key)
+    {
+        return content.remove(key) != null;
+    }
+
+    public boolean removeChunk(World world, ChunkPos pos, boolean notify)
+    {
+        if (notify)
+        {
+            for (TileEntity te : getChunkTEList(pos))
+            {
+                te.onChunkUnload();
+            }
+        }
+
+        return removeChunk(pos);
+    }
+
+    public int totalCount()
+    {
+        int size = 0;
+        for (List<TileEntity> values : content.values())
+        {
+            size += values.size();
+        }
+        return size;
+    }
+
+    public int chunkCount()
+    {
+        return content.size();
+    }
+
+    @Override
+    public void clear()
+    {
+        content.clear();
+    }
+
+    @Override
+    public String toString()
+    {
+        return content.toString();
+    }
+
+    @Override
+    public boolean isEmpty()
+    {
+        return chunkCount() == 0;
+    }
+
+    @Override
+    public Iterator<TileEntity> iterator()
```

Comment:
I guess you keep an iterable like this: https://github.com/google/guava/blob/master/guava/src/com/google/common/collect/Iterables.java#L499-L500
and you return the iterator of that iterable every time the iterator is called